numpy/core/src/multiarray/common.c

```diff
@@ -162,19 +162,16 @@ PyArray_DTypeFromObjectHelper(PyObject *obj, int maxdims,
             PyObject *temp;
 
             if (string_type == NPY_STRING) {
-                if ((temp = PyObject_Str(obj)) == NULL) {
+                if ((temp = PyObject_Bytes(obj)) == NULL) {
                     goto fail;
                 }
-                itemsize = PyUnicode_GetLength(temp);
+                itemsize = PyBytes_Size(temp);
             }
```

Comment:
This looked plain wrong, so I changed it. No idea how this code is hit

Reply:
Do you mean S3 or U3 here?

Reply:
Sorry, use `b"as"` above, and it will be "S3" not U3.

Reply:
Yeah, you're right - with this change the resulting array is type objects instead of bytes.

Arguably the current behavior is nonsense because coercion of arbitrary objects to bytes is a python 2 thing, python 3 should be pushing the user to pick encodings.

I'll roll back these lines.

Reply:
Rolled back this change, and added some comments to make the original intent a little clearer.

Reply:
Thanks, I obviously agree that this only makes sense on Python 2, but OTOH, the switch is much less clear in the NumPy world...